File: examples/Dockerfile.go
# Go application with good practices but a few issues
FROM golang:1.21-alpine AS builder

WORKDIR /app

# Copy go.mod and go.sum first for better caching
COPY go.mod go.sum ./
RUN go mod download

# Copy source code
COPY . .

# Build the application
RUN CGO_ENABLED=0 GOOS=linux go build -o app .

# Use scratch for smallest possible image
FROM scratch

WORKDIR /app

# Copy binary from builder stage
COPY --from=builder /app/app .

# Add CA certificates for HTTPS
COPY --from=builder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/

# Expose port - but no usage explanation
EXPOSE 8080

# No HEALTHCHECK (not easy in scratch)
# No USER specified (but scratch doesn't have users)

# Run the application
CMD ["/app/app"]

File: examples/Dockerfile.java
# Java application with multistage build but other issues
FROM openjdk:17 AS builder

WORKDIR /build

# Copy pom first for better caching
COPY pom.xml .
# Use wildcard copy for src
COPY src ./src

# Build with Maven
RUN ./mvnw package -DskipTests

# Second stage but still using large base image
FROM openjdk:17

WORKDIR /app

# Copy JAR from builder stage
COPY --from=builder /build/target/*.jar app.jar

# Expose port
EXPOSE 8080

# Set healthcheck
HEALTHCHECK --interval=30s --timeout=3s \
  CMD curl -f http://localhost:8080/actuator/health || exit 1

# Still running as root (no USER)

# Run application
CMD ["java", "-jar", "app.jar"]